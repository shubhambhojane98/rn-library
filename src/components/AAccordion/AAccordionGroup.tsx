import * as React from 'react';

type Props = {
  /**
   * Function to execute on selection change.
   */
  onAccordionPress?: (expandedId: string | number) => void;
  /**
   * Id of the currently expanded list accordion
   */
  expandedId?: string | number;
  /**
   * React elements containing list accordions
   */
  children: any;
};

export type ListAccordionGroupContextType = {
  expandedId: string | number | undefined;
  onAccordionPress: (expandedId: string | number) => void;
} | null;

export const ListAccordionGroupContext =
  React.createContext<ListAccordionGroupContextType>(null);

const ListAccordionGroup = ({
  expandedId: expandedIdProp,
  onAccordionPress,
  children,
}: Props) => {
  const [expandedId, setExpandedId] = React.useState<
    string | number | undefined
  >(undefined);

  const onAccordionPressDefault = (newExpandedId: string | number) => {
    setExpandedId(currentExpandedId =>
      currentExpandedId === newExpandedId ? undefined : newExpandedId,
    );
  };

  return (
    <ListAccordionGroupContext.Provider
      value={{
        expandedId: expandedIdProp || expandedId, // component can be controlled or uncontrolled
        onAccordionPress: onAccordionPress || onAccordionPressDefault,
      }}>
      {children}
    </ListAccordionGroupContext.Provider>
  );
};

ListAccordionGroup.displayName = 'List.AccordionGroup';

export default ListAccordionGroup;
