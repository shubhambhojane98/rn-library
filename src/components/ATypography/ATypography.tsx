import React from 'react';
import {StyleProp, StyleSheet, Text, TextProps, TextStyle} from 'react-native';
import {moderateScale} from 'react-native-size-matters';
import {fontFamilies} from '../../constants/fontFamilies';
import {TextAlignment, TypographyVariant} from './ATypographyEnum';
import {defaultScale} from '../../utils/Common';
import {withTheme, useTheme} from '../../core/theming';
import type {Theme} from '../../utils/types';
interface OwnProps {
  variant?: TypographyVariant;
  color?: string;
  style?: StyleProp<TextStyle>;
  textAlign?: TextAlignment;
  fontSize?: number;
  theme: Theme;
}

const defaultFontSize = moderateScale(14, defaultScale);

type Props = OwnProps & TextProps;
const Typography: React.FC<Props> = ({
  variant = TypographyVariant.PRIMARY,
  color,
  style,
  textAlign = TextAlignment.Left,
  children,
  fontSize = defaultFontSize,
  ...otherProps
}) => {
  const {colors} = useTheme();
  return (
    <Text
      style={[
        styles({
          fontSize,
        })[variant],
        {color: color ? color : colors.textColor},
        {textAlign},
        style,
      ]}
      {...otherProps}>
      {children}
    </Text>
  );
};

export default withTheme(Typography);

const styles = (props: {fontSize: number}) =>
  StyleSheet.create({
    primary: {
      fontFamily: `${fontFamilies.Fonts.primary}-Regular`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: 'normal',
    },
    primarySemiBold: {
      fontFamily: `${fontFamilies.Fonts.primary}-SemiBold`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: '600',
    },
    primaryBold: {
      fontFamily: `${fontFamilies.Fonts.primary}-Bold`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: 'bold',
    },
    secondary: {
      fontFamily: `${fontFamilies.Fonts.secondary}-Regular`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: 'normal',
    },
    secondaryDemi: {
      fontFamily: `${fontFamilies.Fonts.secondary}-Demi`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: '600',
    },
    secondaryBold: {
      fontFamily: `${fontFamilies.Fonts.secondary}-Bold`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: 'bold',
    },
  });
