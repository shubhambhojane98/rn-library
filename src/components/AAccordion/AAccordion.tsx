import React, {useState} from 'react';
import {
  View,
  TouchableOpacity,
  StyleSheet,
  LayoutAnimation,
  Platform,
  UIManager,
} from 'react-native';
import ATypography from '../ATypography/ATypography';
import {TypographyVariant} from '../ATypography/ATypographyEnum';
import {Color} from '../../theme';
import IconSVG from '../../assets/svgs';

interface AccourdionItem {
  id: string | number;
  title: string;
  component: any;
}

interface Prop {
  data: AccourdionItem[];
  marginTop?: number;
  marginBottom?: number;
  marginLeft?: number;
  marginRight?: number;
  separatorColor?: string;
  fontSize?: number;
  color?: string;
}

const defaultMargin = 0;

const AAccordion: React.FC<Prop> = ({
  data,
  marginTop = defaultMargin,
  marginBottom = defaultMargin,
  marginLeft = defaultMargin,
  marginRight = defaultMargin,
  separatorColor = Color.greyOpac20,
  fontSize = 16,
  color = Color.black,
}) => {
  if (Platform.OS === 'android') {
    UIManager.setLayoutAnimationEnabledExperimental(true);
  }
  const [expanded, setExpanded] = useState(false);
  const [selectedItem, setSelectedItem] = useState({id: ''});
  const [icon, SetIcon] = useState('downarrow');

  const stylesWithProps = styles({
    marginTop,
    marginRight,
    marginBottom,
    marginLeft,
    separatorColor,
  });

  const toggleExpand = (item: any) => {
    LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
    setExpanded(true);
    setSelectedItem(item);
  };

  return (
    <View>
      {data.map(item => {
        return (
          <View key={item.id}>
            <TouchableOpacity
              style={stylesWithProps.row}
              onPress={() => toggleExpand(item)}
              activeOpacity={1}>
              <ATypography
                children={item.title}
                fontSize={fontSize}
                color={color}
                variant={TypographyVariant.PRIMARY_BOLD}
                style={stylesWithProps.titleStyle}
              />
              <IconSVG
                name={selectedItem.id === item.id ? 'uparrow' : 'downarrow'}
                height={20}
                width={20}
              />
            </TouchableOpacity>
            <View style={stylesWithProps.separator} />
            {selectedItem.id === item.id && (
              <View style={stylesWithProps.child}>{item.component}</View>
            )}
          </View>
        );
      })}
    </View>
  );
};

export default AAccordion;

const styles = (props: {
  marginTop?: number;
  marginRight?: number;
  marginBottom?: number;
  marginLeft?: number;
  separatorColor?: string;
}) =>
  StyleSheet.create({
    title: {
      fontSize: 12,
      fontFamily: 'DMSans-Medium',
      color: Color.grey,
      width: '80%',
    },
    row: {
      flex: 1,
      flexDirection: 'row',
      justifyContent: 'space-between',
      minHeight: 26,
      paddingVertical: 12,
      alignItems: 'center',
      backgroundColor: 'transparent',
      marginTop: props.marginTop,
      marginRight: props.marginRight,
      marginBottom: props.marginBottom,
      marginLeft: props.marginLeft,
    },
    titleStyle: {flex: 0.95},
    separator: {height: 1, backgroundColor: props.separatorColor},
    child: {
      backgroundColor: 'transparent',
      padding: 0,
      marginTop: props.marginTop,
      marginRight: props.marginRight,
      marginBottom: props.marginBottom,
      marginLeft: props.marginLeft,
    },
  });
