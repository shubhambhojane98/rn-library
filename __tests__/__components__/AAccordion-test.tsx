import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import AAccordion from '../../src/components/AAccordion/AAccordion';
import ATypography from '../../src/components/ATypography/ATypography';

const listData = [
  {
    id: '1',
    title: 'Lorem ipsum dolor sit amet',
    component: (
      <ATypography>
        Consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore
        et dolore magna aliqua. Ut enim ad minim veniam.
      </ATypography>
    ),
  },
  {
    id: '2',
    title: 'Sed do eiusmod tempor incididunt ut labore',
    component: (
      <ATypography>
        Adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore
        magna aliqua. Ut enim ad minim.
      </ATypography>
    ),
  },
  {
    id: '3',
    title: 'Lorem enim ad minim veniam',
    component: (
      <ATypography>
        Do eiusmod tempor incididunt ut labor dolore magna aliqua. Ut enim minim
        veniam.
      </ATypography>
    ),
  },
];

it('renders correctly with default props', () => {
  const tree = renderer.create(<AAccordion data={listData} />).toJSON();
  expect(tree).toMatchSnapshot();
});
