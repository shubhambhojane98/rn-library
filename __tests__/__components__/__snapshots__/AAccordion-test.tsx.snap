// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with default props 1`] = `
<View>
  <View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "minHeight": 26,
          "opacity": 1,
          "paddingVertical": 12,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Mulish-Bold",
              "fontSize": 19.65714285714286,
              "fontWeight": "bold",
            },
            Object {
              "color": "#231F20",
            },
            Object {
              "textAlign": "left",
            },
            Object {
              "flex": 0.95,
            },
          ]
        }
      >
        Lorem ipsum dolor sit amet
      </Text>
      <SvgMock
        name="downarrow"
      />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(53, 53, 53, 0.2)",
          "height": 1,
        }
      }
    />
  </View>
  <View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "minHeight": 26,
          "opacity": 1,
          "paddingVertical": 12,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Mulish-Bold",
              "fontSize": 19.65714285714286,
              "fontWeight": "bold",
            },
            Object {
              "color": "#231F20",
            },
            Object {
              "textAlign": "left",
            },
            Object {
              "flex": 0.95,
            },
          ]
        }
      >
        Sed do eiusmod tempor incididunt ut labore
      </Text>
      <SvgMock
        name="downarrow"
      />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(53, 53, 53, 0.2)",
          "height": 1,
        }
      }
    />
  </View>
  <View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "minHeight": 26,
          "opacity": 1,
          "paddingVertical": 12,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Mulish-Bold",
              "fontSize": 19.65714285714286,
              "fontWeight": "bold",
            },
            Object {
              "color": "#231F20",
            },
            Object {
              "textAlign": "left",
            },
            Object {
              "flex": 0.95,
            },
          ]
        }
      >
        Lorem enim ad minim veniam
      </Text>
      <SvgMock
        name="downarrow"
      />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(53, 53, 53, 0.2)",
          "height": 1,
        }
      }
    />
  </View>
</View>
`;
